### DISCUSSION TITLE
trash-cli on Synology's DSM  OS?

### DISCUSSION THREAD

**Discussion by halloleo** [andreafrancia/trash-cli#349] [Q&A]
Does anybody have experience with installing and using trash-cli on Synology's DSM  OS?

DSM offers a recycle bin, but I but I don't know how compatible it is with trash-cli.

**Comment by andreafrancia** [comment]
I don't know trashbin of Synology's DSM OS but I know the trash-spec [1] that are used by trash-cli and other (Gnome, KDE, ...).

Please enable the DSM trahsbin, trash something, and then run the following commands and, if you want, send the result. 

```
$ df
$ ls /Trash-* -ladR
$ ls /Trash -ladR
$ ls ~/.local/share/Trash/ -ladR
```

If any of the `ls` command prints something DSM probably support the trash-spec [1] and it is compatible with trash-cli.

If the df output seems to indicate that there are more partitions try also these two commands for each top level dir.
```
$ ls $topdir/Trash-* -ladR
$ ls $topdir/Trash -ladR
```

If you can install trash-cli just put something in the DSM trashbin and the use the command `trash-list` see if the trash-cli can detect the trashed file. If it does DSM probably support in part of completely the trash-spec [1].

Another thing you can try use DSM to trash two files with the same name (for example test.txt), and the restore only the second file. If the second file get a name different from the original (e.g. `test_1.txt` or `test (2).txt`) then the DSM is not using the trash-spec.

Let me know the result.

[1] https://specifications.freedesktop.org/trash-spec/1.0/

**Comment by mwshook** [comment]
I didn't post the question, but it seems to me that Synology's recycle bin implementation would be incompatible with trash-cli. Each shared folder has its own recycle bin, called `#recycle` in the DSM GUI. From the command line, you can see a hidden folder called something like `Trash-1000`. Inside is a directory called `files` (with the deleted files) and `info` (with a .trashinfo file for each deleted file).

The format of the `Trash-1000` folders looks the same as trash-spec. But the practice of putting a separate trash folder in each shared folder makes trash-cli incompatible. I imagine coding a Synology-compatible version would be possible but I have no concept of how much work that would entail.

Synology's system of permissions is not the same as as Linux, causing some friction. People with write access to a file might be using Windows/SMB and not have a .local directory. A common trash for a whole volume might cause security issues (snooping through trash that doesn't belong to you).

It's annoying, as I find manipulating files easier from the command line, at times.

Note `#recycle`  (The GUI folder) and `Trash-1000` (the actual files and metadata)
![Screenshot from 2025-03-31 21-52-51](https://github.com/user-attachments/assets/55710d9e-8281-43e0-b5f9-7bf49ca1d049)

**Comment by andreafrancia** [comment]
The .Trash-1000 is probably created by trash-cli running on the machine that mounts "autobookshelf" or "comics" as volume when you use trash-put.

For example if you use:
```
$ trash-put /mnt/media/comics/ciao.txt
```
And the you have enough permission the directory /media/comics/.Trash-1000 will be created.
This directory is created by trash-cli and so fully compliant with trash-cli.

The Free Desktop Trash Can Specification consider three directories when the user ask the software to trash a file.
 - $HOME/.local/share/Trash (that should be used if the trashed file is in the same volume of $HOME/.local/share)
 - $topdir/.Trash/$uid (that can be used the admin already created $topdir/.Trash with some specific permissions)
 - $topdir/.Trash-$uid (that can be used when the software can create it)

$HOME/.local/share/Trash is called "home trash" and, according the specifications, can be used to trash file also from another volumes. trash-cli prefers using trash cans from the volumes containing the file to be trashed. 

In the examples before it used as trash can `/media/comics/.Trash-1000` where:
 - `1000` is the user id `$uid`
 - /mnt/media/comics is the mount point of the Synology's on the machine

In this case the other options  `/media/comics/.Trash/1000` has not be used because there was not a preconfigured `/media/comics/.Trash` directory.

Someone with a Synology should see what happen inside the `#recycle` after trashing some files.

**Reply by mwshook** [reply]
I gave up on this when I realized Synology's heavily customized Linux distro wasn't compatible with the Free Desktop trash spec and probably never will be. At one point I was in the comical situation where "trash empty" would put the files in `#recycle`.
Now, I just have an alias `trs` that moves the files to ` ~/.local/share/Trash/files` 
It might be slightly slower than putting the files in `/media/FOLDER/.Trash-1000` and has no provision for restoring, but it works well enough. I don't think it's actually slower because it's all still on the same physical volume.

As long as Synology uses its nonstandard `#Recycle` folder, I don't this will be fixable by trash-cli.

**Reply by andreafrancia** [reply]
Maybe your are mounting your Synology folders with SAMBA/CIFS and the Synology samba is configured to use this: https://www.samba.org/samba/docs/current/man-html/vfs_recycle.8.html

With this module enabled linux samba server will intercept deletion from the client and instead of deleting files move then in a specified "recycle" dir.

So when you use trash-empty it calls the system call corresponding to rm shell command ([unlink(2)](https://man7.org/linux/man-pages/man2/unlink.2.html)) the `vfs_recycle` active in the Synology Samba moves all files to the #recycle dir.

If your Synology supports NFS and you dare to use it maybe trash-cli starts working as intended.
